### DISCUSSION TITLE
conda install from pyscf channel

### DISCUSSION THREAD

**Discussion by jeanwsr** [pyscf/pyscf#2651] [Q&A] [answered]
Currently `conda install pyscf -c conda-forge` works good but `conda install pyscf -c pyscf` (the way recommended by documentation) gives pyscf 2.2.0 instead of 2.6.1. Is there something wrong with the pyscf channel?

**Comment by chillenb** [comment]
Weird--I can't reproduce this problem. What is the output from `conda info` and `conda list`?

**Comment by jeanwsr** [comment]
```
(pyscf-py39) [testpy@yogalab ~]$ conda info

     active environment : pyscf-py39
    active env location : /home/testpy/miniconda3/envs/pyscf-py39
            shell level : 2
       user config file : /home/testpy/.condarc
 populated config files : /home/testpy/.condarc
          conda version : 23.1.0
    conda-build version : not installed
         python version : 3.7.10.final.0
       virtual packages : __archspec=1=x86_64
                          __glibc=2.39=0
                          __linux=6.6.32=0
                          __unix=0=0
       base environment : /home/testpy/miniconda3  (writable)
      conda av data dir : /home/testpy/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/testpy/miniconda3/pkgs
                          /home/testpy/.conda/pkgs
       envs directories : /home/testpy/miniconda3/envs
                          /home/testpy/.conda/envs
               platform : linux-64
             user-agent : conda/23.1.0 requests/2.28.1 CPython/3.7.10 Linux/6.6.32-1-MANJARO manjaro/24.0.1 glibc/2.39
                UID:GID : 1002:1000
             netrc file : None
           offline mode : False

(pyscf-py39) [testpy@yogalab ~]$ conda list
# packages in environment at /home/testpy/miniconda3/envs/pyscf-py39:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_openmp_mutex             5.1                       1_gnu  
ca-certificates           2024.3.11            h06a4308_0  
ld_impl_linux-64          2.38                 h1181459_1  
libffi                    3.4.4                h6a678d5_1  
libgcc-ng                 11.2.0               h1234567_1  
libgomp                   11.2.0               h1234567_1  
libstdcxx-ng              11.2.0               h1234567_1  
ncurses                   6.4                  h6a678d5_0  
openssl                   3.0.13               h7f8727e_2  
pip                       24.0             py39h06a4308_0  
python                    3.9.19               h955ad1f_1  
readline                  8.2                  h5eee18b_0  
setuptools                69.5.1           py39h06a4308_0  
sqlite                    3.45.3               h5eee18b_0  
tk                        8.6.14               h39e8969_0  
tzdata                    2024a                h04d1e81_0  
wheel                     0.43.0           py39h06a4308_0  
xz                        5.4.6                h5eee18b_1  
zlib                      1.2.13               h5eee18b_1  
```

**Comment by chillenb** [comment]
I think the problem is that the conda-forge channel is not enabled. The official PySCF conda packages depend on libstdc++ and libgcc >= 13.2.0, and those aren't available through Anaconda's main channel.

Try `conda install -c pyscf -c conda-forge pyscf`, or if that fails, `conda create -n testenv123 -c pyscf -c conda-forge pyscf`. You could also enable the conda-forge channel in your $HOME/.condarc.

**Comment by jeanwsr** [comment]
Thanks for your help!

`conda install -c pyscf -c conda-forge pyscf` simply fails.

`conda create -n xxx -c pyscf -c conda-forge python=3.9 pyscf` works. But `conda create -n xxx -c pyscf -c conda-forge python=3.9` then `conda install pyscf -c pyscf` fails. That doesn't matter... at least there's one approach works.

But, if the conda-forge channel has to be enabled, the correct instruction should be described in [documentation](https://pyscf.org/install.html#installation-with-conda), which is currently wrong. Or, just recommend `conda install pyscf -c conda-forge` since it's simple and it works.

By the way, may I ask why pyscf does not support installation with libgcc from default (main) channel (i.e. not packaged with libgcc from default channel) any more? What's the reason for binding to conda-forge channel?

**Comment by chillenb** [comment, marked as answer]
No problem!

I think if you put the conda-forge channel in your .condarc, all those commands would work. More and more installations of conda come with this by default; perhaps that's why the documentation recommends `conda install pyscf -c pyscf`.

As for why compatibility with `main` is not supported; I am not sure. Is there a feature you need that's only offered in main and not conda-forge?

(The libgcc requirement comes from the compiler used to build pyscf--by default, it is gcc 13.2.0 from conda-forge.)

**Comment by jeanwsr** [comment]
I am one of the maintainer of the [MOKIT](gitlab.com/jxzou/mokit)(a MR workflow prog)'s conda package, which is built with gcc 11 from main channel (of course we could switch to conda-forge, but at the beginning I only knew the most common channel and use it -- as pyscf did two or three years ago). Some of our users reported that installing mokit is not compatible with pyscf from conda-forge and pyscf channel so I have to dig into it and ask the question above out of curiosity. Our current solution is to use pyscf from pip and it works good. And switching to conda-forge is another potential choice.

**Comment by chillenb** [comment]
Sorry you are having problems with distributing MOKIT. If you have a moment, could you try running

```
conda create -n mokitenv -c mokit -c pyscf -c conda-forge --override-channels mokit pyscf
```

This worked for me. You can also add the options `... "libblas=*=*mkl" "python=3.11"` to make NumPy use MKL and get a newer version of Python. Hopefully it works since it installs, but I didn't run any tests. In many cases, `conda install` will fail due to unsatisfiable constraints even though `conda create` succeeds.

Breakdown of the command, in case someone else finds this thread:

- `conda create -n mokitenv` -- create a conda environment named `mokitenv`
-  `-c mokit -c pyscf -c conda-forge --override-channels` ... using the channels mokit, pyscf, and conda-forge, with priority in that order, and do not use any other channels
- `mokit pyscf` ... and install the packages mokit and pyscf.

For reference, here's the output of `conda info` on my computer (WSL2):
```
     active environment : None
            shell level : 0
       user config file : /home/chillenb/.condarc
 populated config files : /home/chillenb/miniforge3/.condarc
                          /home/chillenb/.condarc
          conda version : 24.1.2
    conda-build version : not installed
         python version : 3.12.2.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=zen2
                          __conda=24.1.2=0
                          __cuda=12.4=0
                          __glibc=2.35=0
                          __linux=5.15.146.1=0
                          __unix=0=0
       base environment : /home/chillenb/miniforge3  (writable)
      conda av data dir : /home/chillenb/miniforge3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/chillenb/miniforge3/pkgs
                          /home/chillenb/.conda/pkgs
       envs directories : /home/chillenb/miniforge3/envs
                          /home/chillenb/.conda/envs
               platform : linux-64
             user-agent : conda/24.1.2 requests/2.31.0 CPython/3.12.2 Linux/5.15.146.1-microsoft-standard-WSL2 ubuntu/22.04.4 glibc/2.35 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.7
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

**Comment by chillenb** [comment]
(also, [recent conda versions](https://conda.org/blog/2023-11-06-conda-23-10-0-release/) have gotten much faster and it could be worth upgrading)

**Comment by jeanwsr** [comment]
> Sorry you are having problems with distributing MOKIT. If you have a moment, could you try running
> 
> ```
> conda create -n mokitenv -c mokit -c pyscf -c conda-forge --override-channels mokit pyscf
> ```
> 
> This worked for me. You can also add the options `... "libblas=*=*mkl" "python=3.11"` to make NumPy use MKL and get a newer version of Python. Hopefully it works since it installs, but I didn't run any tests. In many cases, `conda install` will fail due to unsatisfiable constraints even though `conda create` succeeds.
> 
> Breakdown of the command, in case someone else finds this thread:
> 
>     * `conda create -n mokitenv` -- create a conda environment named `mokitenv`
> 
>     * `-c mokit -c pyscf -c conda-forge --override-channels` ... using the channels mokit, pyscf, and conda-forge, with priority in that order, and do not use any other channels
> 
>     * `mokit pyscf` ... and install the packages mokit and pyscf.
> 
> 
> For reference, here's the output of `conda info` on my computer (WSL2):
> 
> ```
>      active environment : None
>             shell level : 0
>        user config file : /home/chillenb/.condarc
>  populated config files : /home/chillenb/miniforge3/.condarc
>                           /home/chillenb/.condarc
>           conda version : 24.1.2
>     conda-build version : not installed
>          python version : 3.12.2.final.0
>                  solver : libmamba (default)
>        virtual packages : __archspec=1=zen2
>                           __conda=24.1.2=0
>                           __cuda=12.4=0
>                           __glibc=2.35=0
>                           __linux=5.15.146.1=0
>                           __unix=0=0
>        base environment : /home/chillenb/miniforge3  (writable)
>       conda av data dir : /home/chillenb/miniforge3/etc/conda
>   conda av metadata url : None
>            channel URLs : https://conda.anaconda.org/conda-forge/linux-64
>                           https://conda.anaconda.org/conda-forge/noarch
>           package cache : /home/chillenb/miniforge3/pkgs
>                           /home/chillenb/.conda/pkgs
>        envs directories : /home/chillenb/miniforge3/envs
>                           /home/chillenb/.conda/envs
>                platform : linux-64
>              user-agent : conda/24.1.2 requests/2.31.0 CPython/3.12.2 Linux/5.15.146.1-microsoft-standard-WSL2 ubuntu/22.04.4 glibc/2.35 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.7
>                 UID:GID : 1000:1000
>              netrc file : None
>            offline mode : False
> ```

Thanks, I test this method in several machines and it works!
